### ISSUE TITLE
Update query is slower than expected

### ISSUE THREAD

**Issue by chris-eaton** [zzzprojects/EntityFramework-Extensions#432]
### Description
Hi :) 

Our updates in our current EF6 Services follow a pattern of

```
public async Task Update(T model)
{
  // 1. Retrive object from db based on model.id
  // 2. Update various properties and call save
}
```

I thought this could be much better so turned to `Set.Attach`.  Oddly this seemed slower than the original code.  So I Googled around and found your library 🥇 

However I am still confused as its not performing as expected. I have a Unit Test that performs three tests and validates timings, to prove which is fastest.

In every case the original "pull and update" method is faster, and I am confused as to why.

### Timings

When "Ran" properly

>    Expected elapsedNew to be less than 0.211s and 989.0µs, but found 0.529s and 884.7µs.

When ran in Debug and inspecting the values

- elapsedNew: 0.58
- elapsedAttach 0.77
- elapsedOld: 0.205

### Test

    [IntegrationFact]
    public async Task TestNewUpdateMethod()
    {
        var plotService = new PlotService(this.MockSystemUser);
        var lastPlot = await plotService.GetLatestPlot(141888, PlotMaterialisationOptions.All);

        // Warmup
        lastPlot.PlotValidity = PlotValidity.BadImei;
        await plotService.UpdatePlot(lastPlot);
        await plotService.UpdatePlotAttach(lastPlot);
        await plotService.UpdatePlotOld(lastPlot);

        // EF Extended
        var sw = new Stopwatch();
        sw.Start();
        for (var i = 0; i < 10; i++)
        {
            await plotService.UpdatePlot(lastPlot);
        }
        sw.Stop();
        var elapsedNew = sw.Elapsed;

        // Repo attach
        sw = new Stopwatch();
        sw.Start();
        for (var i = 0; i < 10; i++)
        {
            await plotService.UpdatePlotAttach(lastPlot);
        }
        sw.Stop();
        var elapsedAttach = sw.Elapsed;

        // Original get and set
        sw = new Stopwatch();
        sw.Start();
        for (var i = 0; i < 10; i++)
        {
            await plotService.UpdatePlotOld(lastPlot);
        }
        sw.Stop();
        var elapsedOld = sw.Elapsed;

        var newPlot = await plotService.GetLatestPlot(141888, PlotMaterialisationOptions.All);
        elapsedNew.Should().BeLessThan(elapsedOld);
    }

### Methods

    public async Task UpdatePlotAttach(Plot plot)
    {
        using (var uow = this.UnitOfWorkFactory.Create(auditChanges: false))
        {
            var plotRepo = uow.GetRepository<IPlotRepository>();
            uow.SuspendChangeDetection();

            var dbPlot = new global::Database.Models.Plot()
            {
                Id = plot.Id,
                DriverId = plot.DriverId,
                VehicleId = plot.VehicleId,
                PlotValidity = plot.PlotValidity,
                CompositeLocationId = plot.CompositeLocationId,
                Utc = plot.Utc,
                UtcHost = plot.UtcHost,
                UtcInserted = plot.UtcInserted
            };

            var entry = plotRepo.Attach(dbPlot);
            entry.Property(x => x.DriverId).IsModified = true;
            entry.Property(x => x.VehicleId).IsModified = true;
            entry.Property(x => x.PlotValidity).IsModified = true;
            entry.Property(x => x.CompositeLocationId).IsModified = true;

            uow.ResumeChangeDetection();
            (uow.Context as DbContext).Configuration.ValidateOnSaveEnabled = false;

            await uow.SaveChangesAsync("Plot Updated").EscapeContext();
        }
    }

    public async Task UpdatePlot(Plot plot)
    {
        using (var uow = this.UnitOfWorkFactory.Create(auditChanges: false))
        {
            var plotRepo = uow.GetRepository<IPlotRepository>();

            await plotRepo.Set
                .Where(p => p.Id == plot.Id)
                .UpdateFromQueryAsync(p => new global::Database.Models.Plot()
                {
                    DriverId = plot.DriverId,
                    VehicleId = plot.VehicleId,
                    PlotValidity = plot.PlotValidity,
                    CompositeLocationId = plot.CompositeLocationId
                })
                .EscapeContext();
        }
    }

    public async Task UpdatePlotOld(Plot plot)
    {
        using (var uow = this.UnitOfWorkFactory.Create(auditChanges: false))
        {
            var plotRepo = uow.GetRepository<IPlotRepository>();

            var existingPlot = await plotRepo.GetPlot(plot.Id).FirstOrDefaultAsync().EscapeContext();

            if (existingPlot == null)
            {
                return;
            }

            existingPlot.DriverId = plot.DriverId;
            existingPlot.VehicleId = plot.VehicleId;
            existingPlot.PlotValidity = plot.PlotValidity;
            existingPlot.CompositeLocationId = plot.CompositeLocationId;

            await uow.SaveChangesAsync("Plot Updated").EscapeContext();
        }
    }

**Comment by chris-eaton** [comment]
I realised that I perhaps didnt do the most conclusive test above, so ran all three methods through BenchmarkDoNet.  Here is the result

|           Method |     Mean |     Error |    StdDev | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
|----------------- |---------:|----------:|----------:|------------:|------------:|------------:|--------------------:|
|       UpdatePlot | 40.60 ms | 0.8094 ms | 1.7074 ms |           - |           - |           - |              304 KB |
|    UpdatePlotOld | 18.40 ms | 0.3679 ms | 0.6247 ms |           - |           - |           - |           160.87 KB |
| UpdatePlotAttach | 53.01 ms | 1.0561 ms | 1.4456 ms |           - |           - |           - |              152 KB |

This really highlighted how much faster the old method is, and interestingly, how much less memory it uses too.

Am I being stupid somewhere?

**Comment by chris-eaton** [comment]
I changed my logic from `UpdateQueryAsync` to `UpdateAsync` and it performs much better! Still not quite comparable though.  Note for any observers the drop in memory usage is due to opening the DB Connection manually. Ill paste the code below in case anyone is interested

|           Method |     Mean |    Error |   StdDev |   Median | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
|----------------- |---------:|---------:|---------:|---------:|------------:|------------:|------------:|--------------------:|
|       UpdatePlot | 25.30 ms | 1.029 ms | 3.035 ms | 24.56 ms |           - |           - |           - |           232.89 KB |
|    UpdatePlotOld | 19.19 ms | 1.465 ms | 4.083 ms | 17.29 ms |           - |           - |           - |              120 KB |
| UpdatePlotAttach | 53.92 ms | 1.078 ms | 2.177 ms | 53.67 ms |           - |           - |           - |              168 KB |

```
if (autoOpenConnection)
{
    (context as DbContext).Database.Connection.Open();
}
```

**Comment by JonathanMagnan** [comment]
Hello @chris-eaton ,

## UpdateFromQuery
The goal of the `UpdateFromQueryAsync` method is to create a SQL and make directly an update in the database. The purpose of this method is to update multiple rows at once with the same value, and not a single row if you are looking for performance.

The SQL generated is a little bit more complex than a normal `UPDATE`, so this is expected to be slower with one row but way faster in multiple rows.

The biggest advantage of this method is that you don't have to load your entities to update them as you did with the `UpdatePlotOld`.

## BulkUpdate

If you want performance, you want to perform your update in bulk operations if you have many entities.

```csharp
public async Task UpdatePlotBulkUpdate(List<Plot> plots)
{
    using (var uow = this.UnitOfWorkFactory.Create(auditChanges: false))
    {
		var plotRepo = uow.GetRepository<IPlotRepository>();
		
		await plotRepo.BulkUpdateAsync(plots) .EscapeContext();
    }
}
```

So in this case, you don't call the method entity by entity but instead, pass your plot list.

## BulkSaveChanges

Another way is using the `BulkSaveChanges` method, but again to make some optimization, you need to pass a list and not a single `Plot`:

```csharp
public async Task UpdatePlotOldBulkSaveChanges(List<Plot> plots)
{
    using (var uow = this.UnitOfWorkFactory.Create(auditChanges: false))
    {
        var plotRepo = uow.GetRepository<IPlotRepository>();

		// `Where` could have been improved with `BulkRead` if thousands of ids: https://entityframework-extensions.net/bulk-read
		var plotIds = plots.Select(x => x.Id).ToList();
        var existingPlots = await plotRepo.Where(x => plotIds.Contains(x => x.Id)).ToList();
		var dictExistingPlots = existingPlots.ToDictionary(x => x.Id);
		
		foreach(var plot in plots)
		{
			if(dictExistingPlots.TryGetValue(plot.Id, out var existingPlot)
			{
				existingPlot.DriverId = plot.DriverId;
				existingPlot.VehicleId = plot.VehicleId;
				existingPlot.PlotValidity = plot.PlotValidity;
				existingPlot.CompositeLocationId = plot.CompositeLocationId;
			}
		}

        await uow.BulkSaveChangesAsync().EscapeContext();
    }
}
```

> Note: the code has not been tested so it might contain some syntax error

## BatchSaveChanges

Similar to the `BulkSaveChanges` but will be faster in `EF6` when the number of entities is low (100 entities or less for example). Instead of using some `heavy` operation, it will batch all commands generated by Entity Framework but in fewer commands.

```csharp
await uow.BatchSaveChangesAsync("Plot Updated").EscapeContext();
```

---

In short, if you wish to have some performance gains, the goal should be more oriented to understand how to update multiples rows at once as a single update does 1 database roundtrip and that's what you want to avoid.

Let me know if that helped you to discover what alternative technic you could use to get better performance.

Best Regards,

Jon

**Comment by chris-eaton** [comment]
Hi Jon

Thank you for a detailed response.

So, to clarify, the methods provided are more geared towards updating multiple rows at once as opposed to just one? For one row should I avoid your library?

Thanks
Chris

**Comment by JonathanMagnan** [comment]
For one row, you are probably better with SaveChanges unless you want to use some kind of option for a special scenario.

**Comment by chris-eaton** [comment]
Hi @JonathanMagnan, do you mean the normal EF SaveChanges model or is there a version I should use within your library?

Thanks

**Comment by JonathanMagnan** [comment]
Hello @chris-eaton ,

If what you are looking for is performance and need to save row by row like the way you show in your benchmark, I recommend for your case using `SaveChanges`.

If performance in your case is critical, I even recommend you don't even use `Entity Framework` at all for a few scenarios (usually an application only have a few save that are critical, the rest can be slightly slower). By writing a pure SQL `UPDATE` statement, you will surely be even faster as you will skip many validations or additional work Entity Framework needs to do. In addition, you will not need to retrieve the data first before updating it which will be even faster.

Our library is focused to improve performance, but not on improving the performance for only 1 row. The only advantage our library can have here is allowing some additional options such as if you wish to update only a few columns instead of the whole entity.
